pkg/scheduler/scheduler.go

```diff
@@ -256,6 +265,47 @@ func findFlavorForResource(name corev1.ResourceName, val int64, wUsed map[string
 	return "", 0
 }
 
+func flavorSelector(spec *corev1.PodSpec, allowedKeys sets.String) nodeaffinity.RequiredNodeAffinity {
+	var specCopy corev1.PodSpec
+
+	// Remove affinity constraints with irrelevant keys.
+	if len(spec.NodeSelector) != 0 {
+		specCopy.NodeSelector = map[string]string{}
+		for k, v := range spec.NodeSelector {
+			if allowedKeys.Has(k) {
+				specCopy.NodeSelector[k] = v
+			}
+		}
+	}
+
+	affinity := spec.Affinity
+	if affinity != nil && affinity.NodeAffinity != nil && affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
+		var termsCopy []corev1.NodeSelectorTerm
+		for _, t := range affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+			var expCopy []corev1.NodeSelectorRequirement
+			for _, e := range t.MatchExpressions {
+				if allowedKeys.Has(e.Key) {
+					expCopy = append(expCopy, e)
+				}
+			}
+			if len(expCopy) != 0 {
+				termsCopy = append(termsCopy, corev1.NodeSelectorTerm{MatchExpressions: expCopy})
+			}
+		}
+		if len(termsCopy) != 0 {
+			specCopy.Affinity = &corev1.Affinity{
+				NodeAffinity: &corev1.NodeAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
+						NodeSelectorTerms: termsCopy,
+					},
+				},
+			}
+		}
+
+	}
+	return nodeaffinity.GetRequiredNodeAffinity(&corev1.Pod{Spec: specCopy})
```

Comment:
this function returns a LazyErrorNodeSelector, which was there just for backwards compatibility purposes.

We should use a regular NodeSelector and fail assignment if there are parsing errors.

Reply:
I am replicating the scheduler code, I think we need to be consistent with that since it is what will be used eventually to match the pods against nodes.

Reply:
Sg, please add a comment.